projects/packages/waf/src/class-waf-runner.php

```diff
@@ -366,24 +386,39 @@ public static function deactivate() {
 
 		// Empty the rules entrypoint file.
 		if ( ! $wp_filesystem->put_contents( self::get_waf_file_path( Waf_Rules_Manager::RULES_ENTRYPOINT_FILE ), "<?php\n" ) ) {
-			throw new \Exception( 'Failed to empty rules.php file.' );
+			throw new File_System_Exception( 'Failed to empty rules.php file.' );
 		}
 	}
 
 	/**
 	 * Handle updates to the WAF
+	 *
+	 * @throws Waf_Exception If generating the standalone bootstrap file fails.
+	 * @throws Waf_Exception If generating the rule files fails.
+	 *
+	 * @return void
 	 */
 	public static function update_waf() {
-		Waf_Rules_Manager::update_rules_if_changed();
+		$rules_generated_exception = null;
+
+		try {
+			Waf_Rules_Manager::update_rules_if_changed();
+		} catch ( Waf_Exception $e ) {
+			$rules_generated_exception = $e;
+		}
+
 		// Re-generate the standalone bootstrap file on every update
 		// TODO: We may consider only doing this when the WAF version changes
 		try {
 			( new Waf_Standalone_Bootstrap() )->generate();
-		} catch ( \Exception $e ) {
-			return new WP_Error( 'waf_update_failed', $e->getMessage() );
-		}
 
-		return true;
+			// Throw the exception from generating the rules after we've generated the standalone bootstrap file.
+			if ( $rules_generated_exception ) {
+				throw $rules_generated_exception;
+			}
```

Comment:
Do we indeed want to update the bootstrap even though the rule update changed? I wonder in what scenarios that may be prudent, because if the update did not go through, shouldn't the existing bootstrap still be good?

Reply:
I think this is unnecessary - my intention was to ensure the bootstrap file was regenerated in case the autoloader path needed to be updated. But, this method is only used by the update API endpoint to update the rule options... so it isn't really relevant here 😅

I'll remove it :+1:

Reply:
https://github.com/Automattic/jetpack/pull/29108/commits/0c251fe3be28a9024dc0bb75bc44455dfc73bc65 :+1: :+1: